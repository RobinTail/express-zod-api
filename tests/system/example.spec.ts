import fetch from 'node-fetch';
import {spawn, ChildProcessWithoutNullStreams} from 'child_process';

describe('Example', () => {
  let example: ChildProcessWithoutNullStreams;
  let out = '';
  const listener = (chunk: Buffer) => {
    out += chunk.toString();
  };

  beforeAll(() => {
    example = spawn('ts-node', ['example/index.ts']);
    example.stdout.on('data', listener);
  });

  afterAll(async () => {
    example.stdout.removeListener('data', listener);
    example.kill();
    await new Promise((resolve) => {
      const timer = setInterval(() => {
        if (example.killed) {
          clearInterval(timer);
          resolve('OK');
        }
      }, 100);
    });
  });

  afterEach(() => {
    out = '';
  });

  describe('Positive', () => {
    test('Should listen', async () => {
      expect(await new Promise((resolve) => {
        const timer = setInterval(() => {
          if (out.match(/Listening 8090/)) {
            clearInterval(timer);
            resolve('OK');
          }
        }, 100);
      })).toBeTruthy();
    });

    test('Should handle OPTIONS request', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'OPTIONS',
      });
      expect(response.status).toBe(200);
      const text = await response.text();
      expect(text).toBe('');
      expect(response.headers).toBeTruthy();
      expect(response.headers.has('access-control-allow-origin')).toBeTruthy();
      expect(response.headers.has('access-control-allow-methods')).toBeTruthy();
      expect(response.headers.has('access-control-allow-headers')).toBeTruthy();
      expect(response.headers.get('access-control-allow-origin')).toBe('*');
      expect(response.headers.get('access-control-allow-methods')).toBe('POST, OPTIONS');
      expect(response.headers.get('access-control-allow-headers')).toBe('content-type');
    });

    test('Should handle valid POST request', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'POST',
        headers: {
          token: '456',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          key: '123',
          id: 50,
          name: 'John Doe'
        }),
      });
      expect(response.status).toBe(200);
      const json = await response.json();
      expect(json).toEqual({
        status: 'success',
        data: {
          status: "I'll fix it later"
        }
      });
      expect(out).toMatch(/v1\/setUser/);
      expect(out).toMatch(/50, 123, 456/);
    });

    test('Should handle valid GET request', async () => {
      const response = await fetch('http://localhost:8090/v1/getUser?test=123&id=50');
      expect(response.status).toBe(200);
      const json = await response.json();
      expect(json).toEqual({
        status: 'success',
        data: {
          status: 'Some kind of warning',
          name: 'John Doe'
        }
      });
      expect(out).toMatch(/v1\/getUser/);
      expect(out).toMatch(/50, method get/);
    });
  });

  describe('Negative', () => {
    test('GET request should fail on missing input param', async () => {
      const response = await fetch('http://localhost:8090/v1/getUser?test=123');
      expect(response.status).toBe(400);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'id: Required'
        }
      });
    });

    test('GET request should fail on specific value in handler implementation', async () => {
      const response = await fetch('http://localhost:8090/v1/getUser?test=123&id=101');
      expect(response.status).toBe(404);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'User not found'
        }
      });
      expect(out).toMatch(/101, method get/);
    });

    test('POST request should fail on auth middleware key check', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'POST',
        headers: {
          token: '456',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          key: '456',
          id: 50,
          name: 'John Doe'
        }),
      });
      expect(response.status).toBe(401);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'Invalid key'
        }
      });
    });

    test('POST request should fail on auth middleware token check', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'POST',
        headers: {
          token: '123',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          key: '123',
          id: 50,
          name: 'John Doe'
        }),
      });
      expect(response.status).toBe(401);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'Invalid token'
        }
      });
    });

    test('POST request should fail on schema validation', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'POST',
        headers: {
          token: '456',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          key: '123',
          id: -50,
          name: 'John Doe'
        }),
      });
      expect(response.status).toBe(400);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'id: Value should be greater than or equal to 0'
        }
      });
    });

    test('POST request should fail on specific value in handler implementation', async () => {
      const response = await fetch('http://localhost:8090/v1/setUser?test=123&id=50', {
        method: 'POST',
        headers: {
          token: '456',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          key: '123',
          id: 101,
          name: 'John Doe'
        }),
      });
      expect(response.status).toBe(404);
      const json = await response.json();
      expect(json).toEqual({
        status: 'error',
        error: {
          message: 'User not found'
        }
      });
    });
  });
});
