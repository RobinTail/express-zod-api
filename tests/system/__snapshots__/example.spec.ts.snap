// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example Client Should perform the request with a positive response 1`] = `
{
  "data": {
    "features": [
      {
        "features": [
          {
            "features": [],
            "title": "Above 180cm",
          },
        ],
        "title": "Tall",
      },
      {
        "features": [],
        "title": "Young",
      },
      {
        "features": [
          {
            "features": [
              {
                "features": [],
                "title": "About Typescript",
              },
            ],
            "title": "Tells funny jokes",
          },
        ],
        "title": "Cute",
      },
    ],
    "id": 10,
    "name": "John Doe",
  },
  "status": "success",
}
`;

exports[`Example Negative GET request should fail on missing input param 1`] = `
{
  "error": {
    "message": "id: Required",
  },
  "status": "error",
}
`;

exports[`Example Negative POST request should fail on schema validation 1`] = `
{
  "error": {
    "message": "id: should be greater than or equal to 0",
  },
  "status": "error",
}
`;

exports[`Example Negative Should respond with error to the missing static file request 1`] = `
{
  "error": {
    "message": "Can not GET /public/missing.svg",
  },
  "status": "error",
}
`;

exports[`Example Positive Should accept raw data 0 1`] = `
{
  "data": {
    "length": 48687,
  },
  "status": "success",
}
`;

exports[`Example Positive Should accept raw data 1 1`] = `
{
  "data": {
    "length": 48687,
  },
  "status": "success",
}
`;

exports[`Example Positive Should send an image with a correct header 1`] = `"f39beeff92379dc935586d726211c2620be6f879"`;

exports[`Example Positive Should serve static files 1`] = `"f39beeff92379dc935586d726211c2620be6f879"`;

exports[`Example Positive Should stream an image with a correct header 1`] = `"f39beeff92379dc935586d726211c2620be6f879"`;

exports[`Example Positive Should upload the file 1`] = `
{
  "data": {
    "hash": "f39beeff92379dc935586d726211c2620be6f879",
    "mime": "image/svg+xml",
    "name": "logo.svg",
    "otherInputs": {
      "arr": [
        "456",
        "789",
      ],
      "num": "123",
      "obj": {
        "some": "thing",
      },
      "str": "test string value",
    },
    "size": 48687,
  },
  "status": "success",
}
`;
